### DISCUSSION TITLE
Get HTTP status code when user is None

### DISCUSSION THREAD

**Discussion by davidbrochart** [fastapi-users/fastapi-users#730] [Q&A]
Currently, `fastapi_users.current_user(optional=True)` returns `None` "if there is no authenticated user or if it doesn't pass the other requirements", but we loose the HTTP status code that would have been returned if `optional=False` (it can be `401` or `403`).
Do you think it could be possible to retrieve it in e.g. a subsequent call? I was thinking about something like `fastapi_users.get_status_code()`, but I guess it is not guaranteed to refer to our user, since this is all async. Or maybe instead of returning `None`, returning a user with e.g. a `exists=False` attribute?
The reason I'm asking is that I want to nest  `fastapi_users.current_user()` into another dependency, that would just bypass it if no authentication is required (and this configuration is also a dependency):
```python
def my_current_user(optional=False):
    async def _(auth_config=Depends(get_auth_config), user: User = Depends(fastapi_users.current_user(optional=True))):
        if auth_config.no_auth:
            return fake_user
        elif user is None and not optional:
            # raise 401 or 403
        else:
            return user
    return _
```

**Comment by frankie567** [comment]
Hmmm, I see 🤔

The point of the `optional` option was to provide a way to build routes that change their behavior given if there is an authenticated user or not in the request. For example, a "list" endpoint could show only public objects for anonymous requests but authenticated requests will also return private objects.

I didn't think about your use-case. Right now, I don't see any solution other than setting the status codes again yourself. The logic is actually pretty basic:

https://github.com/fastapi-users/fastapi-users/blob/fe79d974ac050cfb55dd87b732ca733b239f60c8/fastapi_users/authentication/__init__.py#L118-L127

---

That said, I'm curious about your use-case here: does `get_auth_config` retrieve a configuration fully dynamically? Is it likely to change for every request? Intuitively, I would say that this is something that could be determined at startup without the need to check it again for each request.

**Reply by davidbrochart** [reply]
But setting the status code again would mean that basically I don't call `fastapi_users.current_user(optional=True)` anymore, because if I do, I only get a `None` user back. Or am I missing something?
For my use-case, having the configuration as a dependency is actually a way to test different configurations, because FastAPI allows to [test dependencies with overrides](https://fastapi.tiangolo.com/advanced/testing-dependencies). The configuration is not likely to change for every request, so this might be overkill, but I don't know a better way to test configuration.

**Reply by frankie567** [reply]
> But setting the status code again would mean that basically I don't call `fastapi_users.current_user(optional=True)` anymore, because if I do, I only get a None user back. Or am I missing something?

What I meant is just to complete your snippet with the relevant status code and exception:

```py
def my_current_user(optional=False):
    async def _(auth_config=Depends(get_auth_config), user: User = Depends(fastapi_users.current_user(optional=True))):
        if auth_config.no_auth:
            return fake_user
        elif user is None and not optional:
            raise HTTPException(status_code=401)
        else:
            return user
    return _
```

--- 

> For my use-case, having the configuration as a dependency is actually a way to test different configurations, because FastAPI allows to test dependencies with overrides. The configuration is not likely to change for every request, so this might be overkill, but I don't know a better way to test configuration.

OK! I get it, then! Well, things will be way simpler then. You can very well do something like this during testing:

```py
get_current_active_user = fastapi_users.current_user(active=True)  # Assign it once and use it across your whole application

# ...

app.dependency_overrides[get_current_active_user] = lambda: fake_user
```

**Reply by davidbrochart** [reply]
> What I meant is just to complete your snippet with the relevant status code and exception

OK but you hard-coded `401`, that's what I meant with the loss of information about the status code (it could be `403`).

> OK! I get it, then! Well, things will be way simpler then. You can very well do something like this during testing

I'm afraid I'm not testing my configuration system if I hard-code the `current_user` dependency like that during testing.

**Reply by frankie567** [reply]
I have to admit I'm not sure to understand what you want to achieve 🤔 Does your configuration system "get_auth_config" has functional implications in your application? Or is it just a trick for testing?

**Reply by davidbrochart** [reply]
Yes the configuration system has functional implications, and it's injected as a dependency to be able to test it easily with dependency overrides. But no worries, I didn't mean to take your time with my special use-case.
What I think though is that it could be interesting to be able to get the status code when `current_user(optional=True)` returns `None`.

**Reply by frankie567** [reply]
I'll think about it!

However, for your configuration system, are you sure that a simple [Pydantic class to manage settings](https://pydantic-docs.helpmanual.io/usage/settings/) would not work for your use-case? You can easily set the value with environment variables or set them at initialization.

That way, settings would be accessible statically instead of passing by a dependency resolved at runtime.

**Reply by davidbrochart** [reply]
Actually using config in a dependency is [recommended by FastAPI](https://fastapi.tiangolo.com/advanced/settings/#settings-in-a-dependency), as it's easily testable. The problem with using a simpler, more static approach is that we have [a plugin system](https://github.com/adriendelsalle/fps) that resolves configuration *after* all routers are initialized.
BTW, just to let you know, we are developing [a new Jupyter server](https://github.com/davidbrochart/jupyverse) based on FastAPI and FastAPI-Users. Thanks for this great project!

**Reply by frankie567** [reply]
That looks great @davidbrochart! 

I've slept on this issue and I think the work I'm currently doing in #729 can solve your problem, in a different way. Basically, my goal is to make everything work as a dependency callable, including database connection ~~and authentication backends. I want to limit the instances and variables living outside the FastAPI lifecycle.~~

EDIT: Maybe I've not slept so well 😅 I'm not sure I'll be able to achieve this without breaking the OpenAPI doc... At some point, FastAPI needs to determine the dependency statically... I'll continue to think about it though 😅

**Comment by frankie567** [comment]
So, I had the chance to think about this, and here is what I would suggest to make your use-case work:

## 1. Implement authentication backend for no-auth

```py
class NoAuthAuthentication(BaseAuthentication):
    def __init__(self, user: models.UserDB, name: str = "noauth"):
        super().__init__(name, logout=False)
        self.user = user

    async def __call__(self, credentials, user_manager):
        # Always return the user no matter what
        return self.user

no_auth_user = models.UserDB(email="foo@bar.com", hashed_password="aaa")
no_auth_authentication = NoAuthAuthentication(no_auth_user)
```

## 2. Implement a dependency returning the authentication backends given the config

```py
async def get_enabled_backends(auth_config = Depends(get_auth_config)):
    if auth_config.mode == "noauth":
        return [no_auth_authentication]
    return [cookie_authentication]
```

## 3. Use the new `get_enabled_backends` argument of `current_user`

This feature was added in [`v8.1.0`](https://github.com/fastapi-users/fastapi-users/releases/tag/v8.1.0). It accepts a parameter `get_enabled_backends` to enable backends dynamically following what's returned by the dependency. So you just have to do:

```py
current_active_user = fastapi_users.current_user(active=True, get_enabled_backends=get_enabled_backends)
```

This way, when in `noauth` mode, you won't get an error because cookie authentication won't be enabled. Therefore, you don't need to tinker with `optional` anymore.

Let me know what you think about this 😉

**Reply by davidbrochart** [reply]
Thanks a lot! I see you are looking at jupyverse's code, feel free to open PRs, I think your expertise would be very valuable there.

**Reply by frankie567** [reply]
I would love to! If you're OK, I can propose the approach I present above 🙂

**Reply by davidbrochart** [reply]
Great, I'm about to open a PR, I'll ping you!
